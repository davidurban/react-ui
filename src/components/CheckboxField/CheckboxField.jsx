import React from 'react';
import UserInputBase from '../UserInputBase';

const CheckboxField = props => (
  <input
    {...props}
    type="checkbox"
  />
);

export default UserInputBase(CheckboxField, true);
