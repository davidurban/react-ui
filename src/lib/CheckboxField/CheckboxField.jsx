import PropTypes from 'prop-types';
import React from 'react';
import styles from './CheckboxField.scss';

const CheckboxField = props => (
  <div className={styles.root}>
    <label htmlFor={props.fieldId} className={styles.inputWrap}>
      <input
        id={props.fieldId}
        name={props.fieldId}
        value={props.value}
        onChange={props.changeHandler}
        disabled={props.disabled}
        checked={props.checked}
        required={props.required}
        type="checkbox"
        className={styles.input}
      />
      <span className={styles.label}>
        <span className={styles.labelInner}>
          {props.label}
        </span>
      </span>
    </label>
    {props.description && (
    <div className={styles.description}>
      {props.description}
    </div>
      )}
    {props.error && (
    <div className={styles.error}>
      {props.error}
    </div>
      )}
  </div>
);

CheckboxField.defaultProps = {
  changeHandler: null,
  checked: false,
  description: null,
  disabled: false,
  error: null,
  required: false,
  value: undefined,
};

CheckboxField.propTypes = {
  changeHandler: PropTypes.func,
  checked: PropTypes.bool,
  description: PropTypes.string,
  disabled: PropTypes.bool,
  error: PropTypes.string,
  fieldId: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  required: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
};

export default CheckboxField;
