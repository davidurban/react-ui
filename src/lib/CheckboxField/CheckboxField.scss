@import '../Theme/mixins';
@import '../Theme/variables';

.root {
  position: relative;
  display: block;
  box-sizing: border-box;
  padding: 0 0.25rem 1.75rem;
}

.inputWrap {
  position: relative;
}

.input {
  @include rui-checkbox-radio-input-styles($rui-checkbox-radio-size, $rui-checkbox-radio-top, 0);
}

.label {
  @include rui-checkbox-radio-label-styles($rui-checkbox-radio-label-height, $rui-checkbox-radio-padding-left);
}

.input + .label::before {
  @include rui-checkbox-radio-shared-inactive(
    $rui-checkbox-radio-size,
    $rui-checkbox-radio-top,
    0,
    $rui-form-field-border-color
  );

  border-radius: $rui-border-radius;
}

.input + .label::after {
  content: '';
  position: absolute;
  top: (1rem - $rui-checkbox-radio-size) / 2;
  left: ($rui-checkbox-checkmark-size / 2);
  display: none;
  width: $rui-checkbox-checkmark-size;
  height: calc((#{$rui-checkbox-checkmark-size} / 2) + 1px);
  border-color: $rui-checkbox-checkmark-color;
  border-bottom: $rui-checkbox-checkmark-border-style;
  border-left: $rui-checkbox-checkmark-border-style;
  transform: translate(0, ($rui-checkbox-checkmark-size / 2)) rotate(-45deg);
}

.input:disabled + .label {
  @include rui-disabled-form-fields();

  color: $rui-checkbox-disabled-font-color;
}

/* stylelint-disable selector-max-specificity */
.input:checked + .label::before {
  border: 0;
  background: $rui-form-field-background-color;
}

.input:checked + .label::after {
  display: inline-block;
}

.input:focus + .label::before {
  outline: 0;
  box-shadow: $rui-radio-focused-box-shadow;
}

.input:focus:checked + .label::before {
  outline: 0;
  box-shadow: $rui-radio-checked-box-shadow, $rui-radio-focused-box-shadow;
}

.input:disabled:checked + .label::before {
  background-color: $rui-radio-disabled-background-color;
  box-shadow: $rui-radio-disabled-checked-box-shadow;
}

.input:required + .label .labelInner::after {
  content: ' *';
}

.input:indeterminate + .label::before {
  border: $rui-border-width solid $rui-checkbox-border-color-indeterminate;
}

.input:indeterminate + .label::after {
  display: inline-block;
  border-bottom-color: $rui-checkbox-border-color-indeterminate;
  border-left: 0;
  transform: translate(0, ($rui-checkbox-checkmark-size / 2));
}

.input:checked:disabled + .label::before {
  background-color: $rui-checkbox-disabled-background-color;
}

.input:checked:disabled + .label::after {
  $disabled-border-style: ($rui-checkbox-radio-size / 8) solid $rui-checkbox-disabled-checkmark-color;

  border-bottom: $disabled-border-style;
  border-left: $disabled-border-style;
}
/* stylelint-enable */

.description,
.error {
  font-size: 0.8rem;
}

.error {
  position: absolute;
  bottom: 0.5rem;
  left: 0.25rem;
  color: $rui-input-invalid-color;
}
