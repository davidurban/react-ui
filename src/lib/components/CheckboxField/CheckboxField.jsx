import PropTypes from 'prop-types';
import React from 'react';
import styles from './CheckboxField.scss';

const CheckboxField = (props) => {
  let labelVisibilityClass = '';
  let labelPositionClass = '';

  if (!props.isLabelVisible) {
    labelVisibilityClass = styles.isLabelHidden;
  }

  if (props.labelPosition === 'before') {
    labelPositionClass = styles.labelPositionBefore;
  } else if (props.labelPosition === 'after') {
    labelPositionClass = styles.labelPositionAfter;
  }

  return (
    <div className={styles.root}>
      <label
        htmlFor={props.fieldId}
        className={(`
          ${styles.inputWrap}
          ${labelPositionClass}
        `).trim()}
      >
        <input
          id={props.fieldId}
          name={props.fieldId}
          value={props.value}
          onChange={props.changeHandler}
          disabled={props.disabled}
          checked={props.checked}
          required={props.required}
          type="checkbox"
          className={styles.input}
        />
        <span className={styles.label}>
          <span
            className={(`
              ${styles.labelInner}
              ${labelVisibilityClass}
            `).trim()}
          >
            {props.label}
          </span>
        </span>
      </label>
      {props.description && (
        <div className={styles.description}>
          {props.description}
        </div>
      )}
      {props.error && (
        <div className={styles.error}>
          {props.error}
        </div>
      )}
    </div>
  );
};

CheckboxField.defaultProps = {
  changeHandler: null,
  checked: false,
  description: null,
  disabled: false,
  error: null,
  isLabelVisible: true,
  labelPosition: 'after',
  required: false,
  value: undefined,
};

CheckboxField.propTypes = {
  changeHandler: PropTypes.func,
  checked: PropTypes.bool,
  description: PropTypes.string,
  disabled: PropTypes.bool,
  error: PropTypes.string,
  fieldId: PropTypes.string.isRequired,
  isLabelVisible: PropTypes.bool,
  label: PropTypes.string.isRequired,
  labelPosition: PropTypes.oneOf(['before', 'after']),
  required: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
};

export default CheckboxField;
