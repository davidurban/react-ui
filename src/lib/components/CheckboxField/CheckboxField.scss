@import '../../shared/settings/borders';
@import '../../shared/settings/forms.deprecated';
@import '../../shared/tools/accessibility';
@import '../../shared/tools/forms.deprecated';

.root {
  position: relative;
  display: flex;
  flex-direction: column;
  box-sizing: border-box;
  padding: 0 0.25rem 0 0;
  margin-bottom: 1rem;
}

.inputWrap {
  position: relative;
}

.input {
  @include form-checkbox-radio-input-styles($checkbox-radio-size, $checkbox-radio-top, 0);
}

.label {
  @include form-checkbox-radio-label-styles($checkbox-radio-label-height, $checkbox-radio-padding-left);
}

.input + .label::before {
  @include form-checkbox-radio-shared-inactive(
    $checkbox-radio-size,
    $checkbox-radio-top,
    0,
    $form-field-border-color
  );

  border-radius: $border-radius;
}

.input + .label::after {
  content: '';
  position: absolute;
  top: (1rem - $checkbox-radio-size) / 2;
  left: ($checkbox-checkmark-size / 2);
  display: none;
  width: $checkbox-checkmark-size;
  height: calc((#{$checkbox-checkmark-size} / 2) + 1px);
  border-color: $checkbox-checkmark-color;
  border-bottom: $checkbox-checkmark-border-style;
  border-left: $checkbox-checkmark-border-style;
  transform: translate(0, ($checkbox-checkmark-size / 2)) rotate(-45deg);
}

// stylelint-disable selector-max-specificity, selector-nested-pattern
.labelPositionBefore {
  .input + .label::before {
    right: 0;
    left: auto;
  }

  .input + .label::after {
    right: ($checkbox-checkmark-size / 2);
    left: auto;
  }

  .label {
    padding-right: $checkbox-radio-padding-left;
    padding-left: 0;
  }
}

.labelPositionAfter {
  .input + .label::before {
    left: 0;
  }

  .input + .label::after {
    right: 0;
  }
}

.input:disabled + .label {
  @include form-disabled-state();

  color: $checkbox-disabled-text-color;
}

.input:checked + .label::before {
  border: 0;
  background: $form-field-background-color;
}

.input:checked + .label::after {
  display: inline-block;
}

.input:indeterminate + .label::before {
  border: $border-width solid $checkbox-border-color-indeterminate;
}

.input:indeterminate + .label::after {
  display: inline-block;
  border-bottom-color: $checkbox-border-color-indeterminate;
  border-left: 0;
  transform: translate(0, ($checkbox-checkmark-size / 2));
}

.input:focus + .label::before {
  outline: 0;
  box-shadow: $radio-focused-box-shadow;
}

.input:focus:checked + .label::before {
  outline: 0;
  box-shadow: $radio-checked-box-shadow, $radio-focused-box-shadow;
}

.input:disabled:checked + .label::before {
  background-color: $radio-disabled-background-color;
  box-shadow: $radio-disabled-checked-box-shadow;
}

.input:required + .label .labelInner::after {
  content: ' *';
}

.input:checked:disabled + .label::before {
  background-color: $checkbox-disabled-background-color;
}

.input:checked:disabled + .label::after {
  $disabled-border-style: ($checkbox-radio-size / 8) solid $checkbox-disabled-checkmark-color;

  border-bottom: $disabled-border-style;
  border-left: $disabled-border-style;
}
// stylelint-enable

.description,
.error {
  font-size: 0.8rem;
}

.error {
  min-height: 1rem;
  color: $form-invalid-color;
}

.isLabelHidden {
  @include accessibility-hide-text();
}
