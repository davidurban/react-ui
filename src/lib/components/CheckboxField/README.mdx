---
name: CheckboxField
menu: 'Inputs'
route: /components/checkbox-field
---

# CheckboxField

CheckboxField allows users to select one or more options from a set.

import {
  Playground,
  Props,
} from 'docz'
import { CheckboxField } from '../..'

## Basic Usage

To implement the CheckboxField component, you need to import it first:

```js
import { CheckboxField } from '@react-ui-org/react-ui';
```

And use it:

<textarea is="docoff-react-preview">
  {() => {
    const [agree, setAgree] = React.useState(true);
    return (
      <CheckboxField
        checked={agree}
        label="I agree"
        onChange={() => setAgree(!agree)}
      />
    );
  }}
</textarea>

See [API](#api) for all available options.

## General Guidelines

- Use the CheckboxField when there are lists of options and the user may
  **select any number of choices,** including zero, one, or several. For
  selecting just a single option from the list, use either the
  [Radio](/components/radio) or [SelectField](/components/select-field)
  component.

- Use the CheckboxField for boolean (true/false) input **in forms that require
  to be confirmed** by a button to become active. To toggle things on or off
  with an immediate effect (without confirmation), use rather the
  [Toggle](/components/toggle) component.

- **Use positive wording for the checkbox label,** so that it's clear what will
  happen when the user turns on the checkbox. Avoid negations such as ‚ÄúDon't
  send me any emails‚Äù which would mean that the user needs to turn the
  checkbox **on** in order for something **not** to happen.

- Only make the CheckboxField's label invisible when there is **another visual
  clue** to guide users through toggling the input.

- When a short label is not enough, use **help texts to guide users** before
  they enter anything.

- Use **clear, calm error messages** when there's a problem with what they
  entered.

- **Visually present groups of choices as groups,** and clearly separate them
  from other groups on the same page.

üìñ [Read more about checkboxes and radios at Nielsen Norman Group.](https://www.nngroup.com/articles/checkboxes-vs-radio-buttons/)

## Help Text

You may provide an additional help text to clarify what will be the impact of
turning the checkbox on or off.

<textarea is="docoff-react-preview">
  {() => {
    const [getNewsletter, setGetNewsletter] = React.useState(true);
    return (
      <CheckboxField
        checked={getNewsletter}
        helpText="We will not bother you more than once a month, we promise!"
        label="Send me newsletter"
        onChange={() => setGetNewsletter(!getNewsletter)}
      />
    );
  }}
</textarea>

## Label Options

In some cases, it may be convenient to visually hide the checkbox label. The
label remains accessible to assistive technologies.

<textarea is="docoff-react-preview">
  {() => {
    const [checked, setChecked] = React.useState(true);
    return (
      <CheckboxField
        checked={checked}
        isLabelVisible={false}
        label="You cannot see this"
        onChange={() => setChecked(!checked)}
      />
    );
  }}
</textarea>

It's also possible to display label before input:

<textarea is="docoff-react-preview">
  {() => {
    const [checked, setChecked] = React.useState(true);
    return (
      <CheckboxField
        checked={checked}
        label="Label precedes input"
        labelPosition="before"
        onChange={() => setChecked(!checked)}
      />
    );
  }}
</textarea>

## States

### Validation States

Validation states visually present the result of validation of the input. You
should always **provide validation message for states other than valid** so
users know what happened and what action they should take or what options they
have.

<textarea is="docoff-react-preview">
  {() => {
    const [agree, setAgree] = React.useState(true);
    return (
      <>
        <CheckboxField
          checked={agree}
          label="I have read and agree with terms and conditions"
          onChange={() => setAgree(!agree)}
          validationState="valid"
        />
        <CheckboxField
          checked={agree}
          label="I have read and agree with terms and conditions"
          onChange={() => setAgree(!agree)}
          validationState="warning"
          validationText="Please wait 10 minutes until we verify your data."
        />
        <CheckboxField
          checked={agree}
          label="I have read and agree with terms and conditions"
          onChange={() => setAgree(!agree)}
          required
          validationState="invalid"
          validationText="You must agree to be able to proceed."
        />
      </>
    );
  }}
</textarea>

### Disabled State

Disabled state makes the input unavailable.

<textarea is="docoff-react-preview">
  <CheckboxField
    disabled
    label="Disabled checkbox"
  />
  <CheckboxField
    checked
    disabled
    label="Disabled checkbox, checked"
  />
</textarea>

## Forwarding HTML Attributes

In addition to the options below in the [component's API](#api) section, you
can specify [React synthetic events] or you can **add whatever HTML attribute
you like.** All attributes that don't interfere with the API are forwarded to
the native HTML `<input>`. This enables making the component interactive and
helps to improve its accessibility.

üëâ Refer to the MDN reference for the full list of supported attributes of the
[checkbox] input type.

## API

<Props table of={CheckboxField} />

## Theming

Head to [Forms Theming](/customize/theming/forms) to see shared form theming
options. On top of that, the following options are available for CheckboxField.

| Custom Property                                                      | Description                                  |
|----------------------------------------------------------------------|----------------------------------------------|
| `--rui-FormField--check__input--checkbox__border-radius`             | Input corner radius                          |
| `--rui-FormField--check__input--checkbox--checked__background-image` | Background image of checked input            |

[React synthetic events]: https://reactjs.org/docs/events.html
[checkbox]: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox#additional_attributes
