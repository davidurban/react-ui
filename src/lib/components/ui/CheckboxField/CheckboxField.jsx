import PropTypes from 'prop-types';
import React from 'react';
import styles from './CheckboxField.scss';

const CheckboxField = (props) => {
  let labelVisibilityClass = '';
  let labelPositionClass = '';
  let rootValidationStateClass = '';

  if (!props.isLabelVisible) {
    labelVisibilityClass = styles.isLabelHidden;
  }

  if (props.labelPosition === 'before') {
    labelPositionClass = styles.labelPositionBefore;
  } else if (props.labelPosition === 'after') {
    labelPositionClass = styles.labelPositionAfter;
  }

  if (props.validationState === 'invalid') {
    rootValidationStateClass = styles.isRootStateInvalid;
  } else if (props.validationState === 'valid') {
    rootValidationStateClass = styles.isRootStateValid;
  } else if (props.validationState === 'warning') {
    rootValidationStateClass = styles.isRootStateWarning;
  }

  return (
    <div
      className={`
        ${styles.root}
        ${rootValidationStateClass}
      `.trim()}
    >
      <label
        htmlFor={props.fieldId}
        className={(`
          ${styles.inputWrap}
          ${labelPositionClass}
        `).trim()}
      >
        <input
          id={props.fieldId}
          name={props.fieldId}
          value={props.value}
          onChange={props.changeHandler}
          disabled={props.disabled}
          checked={props.checked}
          required={props.required}
          type="checkbox"
          className={styles.input}
        />
        <span className={styles.label}>
          <span
            className={(`
              ${styles.labelInner}
              ${labelVisibilityClass}
            `).trim()}
          >
            {props.label}
          </span>
        </span>
      </label>
      {props.description && (
        <div className={styles.description}>
          {props.description}
        </div>
      )}
      {props.error && (
        <div className={styles.error}>
          {props.error}
        </div>
      )}
    </div>
  );
};

CheckboxField.defaultProps = {
  changeHandler: null,
  checked: false,
  description: null,
  disabled: false,
  error: null,
  isLabelVisible: true,
  labelPosition: 'after',
  required: false,
  validationState: null,
  value: undefined,
};

CheckboxField.propTypes = {
  changeHandler: PropTypes.func,
  checked: PropTypes.bool,
  description: PropTypes.string,
  disabled: PropTypes.bool,
  error: PropTypes.string,
  fieldId: PropTypes.string.isRequired,
  isLabelVisible: PropTypes.bool,
  label: PropTypes.string.isRequired,
  labelPosition: PropTypes.oneOf(['before', 'after']),
  required: PropTypes.bool,
  validationState: PropTypes.oneOf(['invalid', 'valid', 'warning']),
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
};

export default CheckboxField;
