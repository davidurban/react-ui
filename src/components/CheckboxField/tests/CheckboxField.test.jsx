import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import CheckboxField from '../index';

describe('rendering', () => {
  it('renders correctly mandatory props only', () => {
    const tree = shallow(<CheckboxField fieldId="test" />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with all props', () => {
    const tree = shallow(<CheckboxField
      label="label"
      disabled
      fieldId="test"
      value="value"
      helpText="some help"
      errors={['some error', 'another error']}
      required
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});
