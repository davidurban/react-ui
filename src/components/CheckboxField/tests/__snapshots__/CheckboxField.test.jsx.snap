// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly mandatory props only 1`] = `
<div>
  <CheckboxField
    disabled={false}
    id="test"
    required={false}
  />
</div>
`;

exports[`rendering renders correctly with all props 1`] = `
<div>
  <CheckboxField
    disabled={true}
    id="test"
    required={true}
    value="value"
  />
  <label
    htmlFor="test"
  >
    label
  </label>
  <div>
    some help
  </div>
  <div
    key="some error"
  >
    some error
  </div>
  <div
    key="another error"
  >
    another error
  </div>
</div>
`;
